Win32CaptureSample/SimpleCapture.cpp

```diff
@@ -71,51 +61,53 @@ void SimpleCapture::Close()
-    Direct3D11CaptureFramePool const& sender,
-    winrt::Windows::Foundation::IInspectable const&)
+bool SimpleCapture::TryResizeSwapChain(const winrt::Windows::Graphics::Capture::Direct3D11CaptureFrame& frame)
 {
-    auto newSize = false;
+    auto const contentSize = frame.ContentSize();
+    if ((contentSize.Width != m_lastSize.Width) ||
+        (contentSize.Height != m_lastSize.Height))
+    {
+        // The thing we have been capturing has changed size, resize the swap chain to match.
+        m_lastSize = contentSize;
+        m_swapChain->ResizeBuffers(2, static_cast<uint32_t>(m_lastSize.Width), static_cast<uint32_t>(m_lastSize.Height),
+            DXGI_FORMAT_B8G8R8A8_UNORM, 0);
+        return true;
+    }
+    return false;
+}
+
+void SimpleCapture::OnFrameArrived(winrt::Direct3D11CaptureFramePool const& sender, winrt::Windows::Foundation::IInspectable const&)
+{
+    auto swapChainResizedToFrame = false;
 
     {
         auto frame = sender.TryGetNextFrame();
+        swapChainResizedToFrame = TryResizeSwapChain(frame);
 
-        if (frame.ContentSize().Width != m_lastSize.Width ||
-            frame.ContentSize().Height != m_lastSize.Height)
-        {
-            // The thing we have been capturing has changed size.
-            // We need to resize our swap chain first, then blit the pixels.
-            // After we do that, retire the frame and then recreate our frame pool.
-            newSize = true;
-            m_lastSize = frame.ContentSize();
-            m_swapChain->ResizeBuffers(
-                2, 
-                (uint32_t)m_lastSize.Width, 
-                (uint32_t)m_lastSize.Height, 
-                static_cast<DXGI_FORMAT>(DirectXPixelFormat::B8G8R8A8UIntNormalized), 
-                0);
-        }
+        winrt::com_ptr<ID3D11Texture2D> backBuffer;
+        winrt::check_hresult(m_swapChain->GetBuffer(0, winrt::guid_of<ID3D11Texture2D>(), backBuffer.put_void()));
+        auto surfaceTexture = GetDXGIInterfaceFromObject<ID3D11Texture2D>(frame.Surface());
+        // copy surfaceTexture to backBuffer
+        m_d3dContext->CopyResource(backBuffer.get(), surfaceTexture.get());
 
+        if (m_captureNextImage)
         {
-            auto frameSurface = GetDXGIInterfaceFromObject<ID3D11Texture2D>(frame.Surface());
-            
-            com_ptr<ID3D11Texture2D> backBuffer;
-            check_hresult(m_swapChain->GetBuffer(0, guid_of<ID3D11Texture2D>(), backBuffer.put_void()));
-
-            m_d3dContext->CopyResource(backBuffer.get(), frameSurface.get());
+            m_captureNextImage = false;
+
+            DirectX::ScratchImage im;
+            winrt::check_hresult(DirectX::CaptureTexture(GetDXGIInterfaceFromObject<ID3D11Device>(m_device).get(),
+                m_d3dContext.get(), surfaceTexture.get(), im));
+            const auto& realImage = *im.GetImage(0, 0, 0);
+            winrt::check_hresult(DirectX::SaveToWICFile(realImage, DirectX::WIC_FLAGS_NONE,
```

Comment:
@robmikh asking for feedback on this code as I will share this with my team

Reply:
Using a GraphicsCaptureItem constructed from a Visual (using `GraphicsCaptureItem::CreateFromVisual`) should be no different from using a GraphicsCaptureItem from the service. The API is designed such that the pipeline should be the same regardless of how you choose to get the item.

This approach works fine (although it the image shows up washed out in the Photos app? It looks fine in Paint and GIMP), but may not be what you want for this scenario. This works great if you want to request snapshots as an auxiliary function to a live or ongoing capture, but can be wasteful if you only want one frame.

I set up a version that wraps the capture API in an IAsyncOperation that returns a IDirect3DSurface which you can then encode in any way you see fit. You can find that in the "saveframe" branch, with the bulk of the code being in `CaptureSnapshot::TakeAsync` and `App::TakeSnapshotAsync` which shows obtaining the frame and encoding it respectively. This version consumes WIC directly, but substituting that with what you've done is probably more robust. I mainly did it so that I have a WIC example to look back at in the future.